client/components/tinymce/plugins/embed-reversal/plugin.js

```diff
@@ -46,36 +43,75 @@ function embedReversal( editor ) {
 			content = content.replace( markup, result.result );
 			editor.fire( 'SetTextAreaContent', { content } );
 		}
-
 		// Trigger an editor change so that dirty detection and autosave
 		// take effect
 		editor.fire( 'change' );
 	}
 
 	function onPaste( event ) {
 		let markup;
+
 		if ( event.clipboardData ) {
 			markup = event.clipboardData.getData( 'text/plain' );
 		} else if ( window.clipboardData ) {
 			markup = window.clipboardData.getData( 'Text' ); // IE11
 		}
 
 		// Check whether pasted content looks like markup
-		if ( ! markup || ! /^<.*>$/.test( markup ) ) {
+		if ( ! maybeMarkup( markup ) ) {
 			return;
 		}
 
-		// If so, queue a request for reversal
-		wpcom
+		if ( isVisualEditMode() ) {
+			// If we're in the visual editor store the markup for use later
+			clipboardData = markup;
+		} else {
+			// If so, queue a request for reversal
+			queueReversal( markup );
+		}
+	}
+
+	function queueReversal( markup ) {
+		return wpcom
 			.undocumented()
 			.site( getSelectedSiteId( store.getState() ) )
 			.embedReversal( markup )
 			.then( partial( replaceMarkup, markup ) )
-			.catch( () => {} );
+			.catch( () => {
+				const element = getWrapperElement();
+				if ( isVisualEditMode() && element ) {
+					for ( const child of element.childNodes ) {
+						element.parentNode.insertBefore( child, element );
+					}
+					editor.dom.remove( element );
+				}
+			} )
+			.finally( () => ( clipboardData = null ) );
+	}
+
+	function getWrapperElement() {
+		return editor.getBody().querySelector( '#' + elementId );
+	}
+
+	function maybeMarkup( markup ) {
```

Comment:
I'm not 100% sure but I feel that `maybe` isn't quite right here as we're returning a boolean. Something like `isMarkup` or `hasMarkup` might be more indicative of it's usage. At a guess if I saw `maybeMarkup` being used I would assume that it's input is either subjected to being marked up, or not, depending on some condition.

Reply:
I agree although the problem is that it's suspected markup rather than definitely is or isn't. This was instead of calling it `looksLikeMarkup` 😄 

I'll stop overthinking it and change it to `isMarkup`